### PR TITLE
Remove non-leaf fields from updater's fieldsets

### PR DESCRIPTION
**Pull request by jennybuckley**
The presence of non-leaf fields in managed fieldsets doesn't actually do anything, it can never result in a conflict, and it can never change the result of an operation. I think we should remove them, since they can ultimately lead to some managers getting stuck in the managedFields list that own nothing but non-leaf fields, and since they never conflict with anything they end up sticking around forever.

This bug does not usually present itself in k8s built-in types, because fields with struct values are not usually nullable, and so they get defaulted to a struct with 0-values for all it's fields, instead of nil, meaning they never appear as 'added' in a compare result. Unfortunately CRs are backed by unstructured, where all 'structs' are actually maps, so we should probably fix this to try to be consistent with builtin types.

### CODE REVIEW

**Comment by jennybuckley** [comment]
rebased

File: typed/typed.go [hunk 1]
```diff
@@ -238,6 +238,8 @@ func merge(lhs, rhs *TypedValue, rule, postRule mergeRule) (*TypedValue, error)
 	mw.typeRef = lhs.typeRef
 	mw.rule = rule
 	mw.postItemHook = postRule
+	mw.postHook = postRule
+	mw.postItemHook = postItemRule
```
Comment (apelisse) [inline typed/typed.go]: You're overriding `postItemHook`

Comment (apelisse) [reply typed/typed.go]: I suspect we could do a better job finding a good name for these variables?

File: typed/typed.go [hunk 2]
```diff
@@ -99,7 +99,7 @@ func (tv TypedValue) ToFieldSet() (*fieldpath.Set, error) {
 // match), or an error will be returned. Validation errors will be returned if
 // the objects don't conform to the schema.
 func (tv TypedValue) Merge(pso *TypedValue) (*TypedValue, error) {
-	return merge(&tv, pso, ruleKeepRHS, nil)
+	return merge(&tv, pso, ruleKeepRHS, nil, nil)
```
Comment (apelisse) [inline typed/typed.go]: The new parameter is nil everywhere?

**Comment by apelisse** [comment]
It's very questionnable that we still need that after Joe's work on atomic <-> associative. I'll close for now, but we can revisit.

**Pull request closed without merging**